src/firebase_functions/pubsub.py

```diff
@@ -32,49 +30,87 @@ class CloudEvent(Generic[T]):
 class Message(Generic[T]):
   message_id: str
   publish_time: dt.datetime
-  data: str
-  attributes: dict[str, str]
-  ordering_key: str
+  data: Optional[str] = None
+  attributes: Optional[dict[str, str]] = None
+  ordering_key: Optional[str] = None
 
   @property
-  def json(self) -> Dict[str, Any]:
-    return {
+  def json(self) -> Optional[T]:
+    try:
+      if self.data is not None:
+        return json.loads(base64.b64decode(self.data).decode('utf-8'))
+      else:
+        return None
+    except Exception as e:
+      raise Exception(
+          f'Unable to parse Pub/Sub message data as JSON: {e}') from e
+
+  def asdict(self) -> dict[str, Any]:
+    dict_message: dict[str, Any] = {
         'message_id': self.message_id,
-        'publish_time': self.publish_time,
         'data': self.data,
-        'attributes': self.attributes,
-        'ordering_key': self.ordering_key,
+        'publish_time': self.publish_time,
     }
 
+    if self.attributes is not None:
+      dict_message['attributes'] = self.attributes
+    if self.ordering_key is not None:
+      dict_message['ordering_key'] = self.ordering_key
 
-CloudEventMessage = CloudEvent[Message[T]]
+    return dict_message
 
 
-class MessagePublishedData(TypedDict):
-  message: object
+@dataclass(frozen=True)
+class MessagePublishedData:
+  message: Message[Union[dict, str, None]]
   subscription: str
 
 
 def pubsub_wrap_handler(
-    func: Callable[[CloudEventMessage], None],
-    raw: dict[str, Any],
-) -> Response:
-  event: CloudEventMessage = CloudEvent(
-      **raw['attributes'],
-      time=dt.datetime.fromisoformat(raw['attributes']['time']),
-      data=Message(
-          **raw['attributes']['data']['message'],
-          publish_time=dt.datetime.fromisoformat(
-              raw['attributes']['data']['message']['publish_time']),
-      ))
-
-  func(event)
+    func: Callable[[CloudEvent[MessagePublishedData]], None],
```

Comment:
```suggestion
    func: Callable[[CloudEvent[MessagePublishedData[T]]], None],
```

Would it not be this?